File: crates/netcfgd-daemon/src/resolv_guard.rs
//! Removing what keeps taking `/etc/resolv.conf` back.
//!
//! [0165](../../../doc/decision/0165-the-file-netcfgd-owns-is-defended.md)
//! made netcfgd notice a foreign write and put its own file back, which is
//! enough to win every round. What it does not do is stop the rounds: a writer
//! that rewrites the file every few seconds leaves the machine's resolver
//! flapping between two answers, and a name looked up in the wrong second gets
//! the wrong server.
//!
//! **This is the last resort and it is deliberately hard to reach.** It fires
//! only where netcfgd was told to own the file outright, and only after the
//! file has been taken away and put back several times in a row -- so a single
//! write during boot, which is ordinary, never reaches it.
//!
//! **It cannot tell who wrote the file, and the consequence is blunt.** There
//! is no way to ask the kernel which process last wrote a path -- `fanotify`
//! could report it and needs `CAP_SYS_ADMIN`, which netcfgd does not take.
//! So what this actually does is signal **every** known resolver-writing
//! program netcfgd did not start, not the one that is interfering. An idle
//! `dhclient` that has written nothing is terminated alongside the one that
//! will not stop.
//!
//! That is defensible and it is not what the instruction sounds like, so it is
//! written here rather than discovered: on a machine where netcfgd has been
//! told to own `resolv.conf`, a foreign DHCP client that is running *will*
//! write that file when its lease renews, so the distinction between "is
//! interfering" and "is going to" is thinner than it looks. It is still a
//! bystander at the moment it is killed, and `tests/live/resolv_defended.sh`
//! asserts that this is what happens rather than letting it be a surprise.
//!
//! **A supervised process is reported and not signalled.** Killing
//! `systemd-resolved` buys seconds: its unit carries `Restart=`, so the
//! supervisor puts it straight back and netcfgd would be fighting something
//! that cannot lose. The answer there is `Conflicts=` in
//! `packaging/systemd/netcfgd-exclusive.conf`, and saying so is worth more
//! than a signal that achieves nothing.

use std::collections::BTreeSet;
use std::path::Path;

/// How many reclaims in a row before netcfgd stops merely rewriting.
///
/// Three rather than one, because one is ordinary. A DHCP client that took a
/// lease, a hook that ran, an operator with an editor -- each writes the file
/// once, netcfgd puts its own back, and nothing else should happen. Something
/// that has done it three times running is not passing through.
pub(crate) const PATIENCE: u32 = 3;

/// Programs known to write `/etc/resolv.conf`.
///
/// **`/proc/<pid>/comm` is truncated to 15 characters**, which is why
/// `systemd-resolved` is spelled `systemd-resolve` here. A name one character
/// too long never matches, and the sweep would report nothing while looking
/// like it had looked -- the vacuous pass in its most literal form.
///
/// `dhcpcd` is on the list and netcfgd starts its own, which is exactly why
/// the pids netcfgd recorded starting are excluded before anything is
/// signalled. Killing its own DHCP client to defend a file that client's lease
/// filled in would be the worst outcome available.
const WRITERS: &[&str] = &[
	"NetworkManager",
	"systemd-resolve",
	"connmand",
	"dhclient",
	"dhcpcd",
	"resolvconf",
];

/// Every pid netcfgd recorded starting, so none of them is a target.
///
/// The backends write `<run>/<kind>/<iface>.pid`, one directory down. A file
/// that cannot be read or does not hold a number is skipped rather than
/// guessed at: the cost of missing one is signalling something netcfgd owns,
/// so the reading is deliberately strict.
fn ours(run_dir: &Path) -> BTreeSet<i32> {
	let mut pids = BTreeSet::new();
	// Own pid first: netcfgd is not called any of the names above, but a
	// future rename should not be able to make it kill itself.
	if let Ok(mine) = i32::try_from(std::process::id()) {
		pids.insert(mine);
	}
	let Ok(kinds) = std::fs::read_dir(run_dir) else {
		return pids;
	};
	for kind in kinds.flatten() {
		let Ok(entries) = std::fs::read_dir(kind.path()) else {
			continue;
		};
		for entry in entries.flatten() {
			if entry.path().extension().is_none_or(|ext| ext != "pid") {
				continue;
			}
			if let Ok(text) = std::fs::read_to_string(entry.path()) {
				if let Ok(pid) = text.trim().parse::<i32>() {
					pids.insert(pid);
				}
			}
		}
	}
	pids
}

/// Signal whatever is taking the file back, and say what was left alone.
///
/// Returns how many processes were signalled, so the caller can say whether
/// the sweep did anything rather than inferring it.
pub(crate) fn sweep(run_dir: &Path) -> usize {
	let mine = ours(run_dir);
	let mut signalled = 0;

	for (pid, program) in netcfgd_sys::process::pids_of_programs(WRITERS) {
		if mine.contains(&pid) {
			// Not a diagnostic anybody needs on every pass: netcfgd's own
			// dhcpcd is on the list by construction and is not interference.
			continue;
		}
		// **Ours by cgroup, which is how netcfgd finds the children it has no
		// record of.** `ours` reads pid files under netcfgd's own run
		// directory, and dhcpcd does not write one there -- it writes
		// `/run/dhcpcd/<iface>-4.pid`. So netcfgd's own DHCP client, and every
		// privsep helper it forks, reached this sweep looking foreign. It
		// survived only because the supervision check below was answering yes
		// about it for the wrong reason; correcting that alone would have made
		// netcfgd terminate the client holding this machine's lease, which
		// this file's own header calls the worst outcome available. 0198.
		if netcfgd_sys::process::in_our_service(pid) {
			continue;
		}
		if !netcfgd_sys::process::shares_network_namespace(pid) {
			// Not a diagnostic: on a machine running containers this is most
			// of what the scan finds, and saying so every three reclaims
			// would bury the lines that matter.
			continue;
		}
		// **Another manager's service, which is not the same as any service.**
		// Everything netcfgd starts inherits netcfgd's own cgroup, so this
		// used to answer yes about netcfgd's own dhcpcd and the sweep declined
		// to signal a process it had started -- leaving in place exactly the
		// interference it exists to remove. Decision 0198.
		if netcfgd_sys::process::is_service_supervised(pid) {
			netcfgd_sys::log_warning!(
				"resolv",
				"{program} (pid {pid}) keeps rewriting resolv.conf and another \
				 service manager restarts it, so netcfgd is not signalling it. A \
				 killed service comes straight back; stand it down with Conflicts= \
				 -- see packaging/systemd/netcfgd-exclusive.conf"
			);
			continue;
		}
		match netcfgd_sys::process::terminate(pid) {
			Ok(()) => {
				eprintln!(
					"netcfgd: terminated {program} (pid {pid}): it took /etc/resolv.conf back {PATIENCE} times running and netcfgd owns that file"
				);
				signalled += 1;
			}
			Err(error) => {
				eprintln!("netcfgd: could not terminate {program} (pid {pid}): {error}");
			}
		}
	}

	if signalled == 0 {
		eprintln!(
			"netcfgd: resolv.conf has been taken back {PATIENCE} times and netcfgd found nothing it could signal"
		);
	}
	signalled
}

File: c/src/daemon/resolv_guard.c
/*
 * resolv_guard.c -- removing what keeps taking `/etc/resolv.conf` back.
 *
 * `daemon.h` carries the whole argument: why this exists, why it is hard to
 * reach, why it cannot tell who wrote the file, and why a supervised process is
 * reported rather than signalled. What is here is the sweep itself, and the one
 * thing worth saying twice is the order of the four exclusions, because each of
 * them was paid for:
 *
 *   1. a pid netcfgd recorded starting, from the backends' pid files;
 *   2. a process in netcfgd's own service, which is how netcfgd finds the
 *      children it has no record of -- dhcpcd writes its pid file to
 *      `/run/dhcpcd/<iface>-4.pid`, where netcfgd does not look, and forks
 *      three helpers that appear in no pid file at all (0198);
 *   3. a process in another network namespace, which is not configuring
 *      netcfgd's interfaces whatever it is called;
 *   4. a process *another* service manager holds up, which is not the same as
 *      any service manager: everything netcfgd starts inherits netcfgd's own
 *      cgroup, so "is in a service" used to answer yes about netcfgd's own
 *      dhcpcd and the sweep declined to signal a process it had started --
 *      leaving in place exactly the interference it exists to remove.
 *
 * The first three are quiet. On a machine running containers the third is most
 * of what the scan finds, and saying so every three reclaims would bury the
 * lines that matter.
 */
#include "ncfg/daemon.h"

#include "ncfg/log.h"
#include "ncfg/process.h"

#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

/*
 * Programs known to write `/etc/resolv.conf`.
 *
 * `systemd-resolved` is spelled `systemd-resolve` because `/proc/<pid>/comm` is
 * truncated to fifteen characters; see `ncfg_resolv_writers`.
 */
static const char *const writers[] = {
	"NetworkManager",
	"systemd-resolve",
	"connmand",
	"dhclient",
	"dhcpcd",
	"resolvconf"
};

const char *const *ncfg_resolv_writers(size_t *count_out)
{
	if (count_out) {
		*count_out = sizeof(writers) / sizeof(writers[0]);
	}
	return writers;
}

/* ------------------------------------------------------------------------ *
 * The pids netcfgd knows are its own
 * ------------------------------------------------------------------------ */

/* Keep `pid`, lowest first and without repeats, and answer the new total. */
static size_t remember(pid_t pid, pid_t *out, size_t out_max, size_t total)
{
	size_t kept = total < out_max ? total : out_max;
	size_t at;
	size_t i;

	for (at = 0; at < kept; at++) {
		if (out[at] == pid) {
			return total;
		}
	}
	/* Sorted, so that overflowing keeps the lowest rather than the first ones
	 * `readdir` happened to hand over. The order is not what makes the answer
	 * right -- the caller refuses to sweep at all when the set did not fit --
	 * but an answer that depends on directory order is one nobody can
	 * reproduce. */
	for (at = 0; at < kept && out[at] < pid; at++) {
	}
	for (i = kept; i > at; i--) {
		if (i < out_max) {
			out[i] = out[i - 1u];
		}
	}
	if (at < out_max) {
		out[at] = pid;
	}
	return total + 1u;
}

/* The number in a pid file, or 0 for anything this cannot vouch for. */
static pid_t pid_in_file(const char *path)
{
	FILE         *file = fopen(path, "r");
	char          line[64];
	char         *end = NULL;
	long          value;
	const char   *at;

	if (!file) {
		return 0;
	}
	if (!fgets(line, sizeof(line), file)) {
		(void)fclose(file);
		return 0;
	}
	(void)fclose(file);
	line[strcspn(line, "\r\n")] = '\0';
	for (at = line; *at == ' ' || *at == '\t'; at++) {
	}
	value = strtol(at, &end, 10);
	/*
	 * Strict on purpose: the cost of missing one of netcfgd's own pids is
	 * signalling something netcfgd started, so a file that cannot be read or
	 * does not hold a number is skipped rather than guessed at.
	 */
	if (end == at || value <= 0 || value > (long)0x7fffffff) {
		return 0;
	}
	while (*end == ' ' || *end == '\t') {
		end++;
	}
	if (*end != '\0') {
		return 0;
	}
	return (pid_t)value;
}

/* Whether a name ends in `.pid`. */
static int is_pid_file(const char *name)
{
	size_t length = strlen(name);

	return length > 4u && strcmp(name + length - 4u, ".pid") == 0;
}

size_t ncfg_resolv_ours(const char *run_dir, pid_t *out, size_t out_max)
{
	DIR    *kinds;
	size_t  total = 0u;

	if (!out) {
		out_max = 0u;
	}
	/*
	 * Own pid first: netcfgd is not called any of the writer names, but a
	 * future rename should not be able to make it kill itself.
	 */
	total = remember(getpid(), out, out_max, total);
	if (!run_dir) {
		return total;
	}
	kinds = opendir(run_dir);
	if (!kinds) {
		return total;
	}
	/*
	 * The backends write `<run>/<kind>/<iface>.pid`, one directory down, so
	 * this walk is exactly two levels and is not a recursion. Nothing here
	 * follows a symlink into a third.
	 */
	for (;;) {
		const struct dirent *kind = readdir(kinds);
		char                 directory[NCFG_CONFIRM_PATH_MAX];
		DIR                 *entries;
		int                  written;

		if (!kind) {
			break;
		}
		if (strcmp(kind->d_name, ".") == 0 || strcmp(kind->d_name, "..") == 0) {
			continue;
		}
		written = snprintf(directory, sizeof(directory), "%s/%s", run_dir, kind->d_name);
		if (written < 0 || (size_t)written >= sizeof(directory)) {
			continue;
		}
		entries = opendir(directory);
		if (!entries) {
			continue;
		}
		for (;;) {
			const struct dirent *entry = readdir(entries);
			char                 path[NCFG_CONFIRM_PATH_MAX];
			pid_t                pid;

			if (!entry) {
				break;
			}
			if (!is_pid_file(entry->d_name)) {
				continue;
			}
			written = snprintf(path, sizeof(path), "%s/%s", directory, entry->d_name);
			if (written < 0 || (size_t)written >= sizeof(path)) {
				continue;
			}
			pid = pid_in_file(path);
			if (pid > 0) {
				total = remember(pid, out, out_max, total);
			}
		}
		(void)closedir(entries);
	}
	(void)closedir(kinds);
	return total;
}

/* ------------------------------------------------------------------------ *
 * The real machine
 * ------------------------------------------------------------------------ */

static size_t real_candidates(void *state, ncfg_process_ref_t *out, size_t out_max)
{
	size_t count = 0u;
	const char *const *names = ncfg_resolv_writers(&count);

	(void)state;
	return ncfg_process_pids_of_programs(names, count, out, out_max);
}

static int real_in_our_service(void *state, pid_t pid)
{
	(void)state;
	return ncfg_process_in_our_service(pid);
}

static int real_shares_network_namespace(void *state, pid_t pid)
{
	(void)state;
	return ncfg_process_shares_network_namespace(pid);
}

static int real_is_service_supervised(void *state, pid_t pid)
{
	(void)state;
	return ncfg_process_is_service_supervised(pid);
}

static int real_terminate(void *state, pid_t pid, char *err, size_t err_size)
{
	(void)state;
	return ncfg_process_terminate(pid, err, err_size);
}

ncfg_resolv_machine_t ncfg_resolv_machine_default(void)
{
	ncfg_resolv_machine_t machine;

	machine.state = NULL;
	machine.candidates = real_candidates;
	machine.in_our_service = real_in_our_service;
	machine.shares_network_namespace = real_shares_network_namespace;
	machine.is_service_supervised = real_is_service_supervised;
	machine.terminate = real_terminate;
	return machine;
}

/* ------------------------------------------------------------------------ *
 * The sweep
 * ------------------------------------------------------------------------ */

static int is_ours(const pid_t *ours, size_t count, pid_t pid)
{
	size_t at;

	for (at = 0; at < count; at++) {
		if (ours[at] == pid) {
			return 1;
		}
	}
	return 0;
}

size_t ncfg_resolv_sweep(const char *run_dir, const ncfg_resolv_machine_t *machine)
{
	pid_t              ours[NCFG_RESOLV_OURS_MAX];
	ncfg_process_ref_t found[NCFG_RESOLV_CANDIDATES_MAX];
	size_t             our_total;
	size_t             found_total;
	size_t             looked_at;
	size_t             at;
	size_t             signalled = 0u;

	if (!machine || !machine->candidates || !machine->in_our_service ||
	    !machine->shares_network_namespace || !machine->is_service_supervised ||
	    !machine->terminate) {
		/* The one place in this library where an incomplete argument means do
		 * nothing rather than fail: what this call does is send signals, and a
		 * caller that has not said which machine has not asked for that. */
		return 0u;
	}
	our_total = ncfg_resolv_ours(run_dir, ours, sizeof(ours) / sizeof(ours[0]));
	if (our_total > sizeof(ours) / sizeof(ours[0])) {
		/*
		 * **Fail closed.** A pid of netcfgd's own that did not fit would look
		 * foreign, and terminating the DHCP client holding this machine's
		 * lease is the worst outcome available.
		 */
		ncfg_log_emitf("resolv", NCFG_LOG_ERROR,
		    "netcfgd has started %lu processes, more than the %d this sweep can account "
		    "for, so nothing will be signalled", (unsigned long)our_total,
		    (int)NCFG_RESOLV_OURS_MAX);
		return 0u;
	}
	found_total = machine->candidates(machine->state, found,
	    sizeof(found) / sizeof(found[0]));
	looked_at = found_total < sizeof(found) / sizeof(found[0])
	    ? found_total
	    : sizeof(found) / sizeof(found[0]);
	if (found_total > looked_at) {
		ncfg_log_emitf("resolv", NCFG_LOG_WARNING,
		    "%lu programs known to write resolv.conf are running and this sweep looks at "
		    "%lu of them", (unsigned long)found_total, (unsigned long)looked_at);
	}

	for (at = 0; at < looked_at; at++) {
		pid_t       pid = found[at].pid;
		const char *program = found[at].program;
		char        why[NCFG_ERROR_MAX];

		/* Not a diagnostic anybody needs on every pass: netcfgd's own dhcpcd
		 * is on the list by construction and is not interference. */
		if (is_ours(ours, our_total, pid)) {
			continue;
		}
		if (machine->in_our_service(machine->state, pid)) {
			continue;
		}
		if (!machine->shares_network_namespace(machine->state, pid)) {
			continue;
		}
		if (machine->is_service_supervised(machine->state, pid)) {
			ncfg_log_emitf("resolv", NCFG_LOG_WARNING,
			    "%s (pid %ld) keeps rewriting resolv.conf and another service manager "
			    "restarts it, so netcfgd is not signalling it. A killed service comes "
			    "straight back; stand it down with Conflicts= -- see "
			    "packaging/systemd/netcfgd-exclusive.conf", program, (long)pid);
			continue;
		}
		why[0] = '\0';
		if (machine->terminate(machine->state, pid, why, sizeof(why))) {
			ncfg_log_emitf("resolv", NCFG_LOG_WARNING,
			    "terminated %s (pid %ld): it took /etc/resolv.conf back %d times "
			    "running and netcfgd owns that file", program, (long)pid,
			    (int)NCFG_RESOLV_PATIENCE);
			signalled++;
		} else {
			ncfg_log_emitf("resolv", NCFG_LOG_ERROR, "could not terminate %s (pid %ld): %s",
			    program, (long)pid, why);
		}
	}

	if (signalled == 0u) {
		ncfg_log_emitf("resolv", NCFG_LOG_WARNING,
		    "resolv.conf has been taken back %d times and netcfgd found nothing it could "
		    "signal", (int)NCFG_RESOLV_PATIENCE);
	}
	return signalled;
}
